src/base/bittorrent/session.cpp

```diff
@@ -442,8 +442,13 @@ Session::Session(QObject *parent)
     , m_networkInterfaceAddress(BITTORRENT_SESSION_KEY("InterfaceAddress"))
     , m_encryption(BITTORRENT_SESSION_KEY("Encryption"), 0)
     , m_isProxyPeerConnectionsEnabled(BITTORRENT_SESSION_KEY("ProxyPeerConnections"), false)
+#if (LIBTORRENT_VERSION_NUM >= 10206)
+    , m_chokingAlgorithm(BITTORRENT_SESSION_KEY("ChokingAlgorithm"), ChokingAlgorithm::RateBased
+        , clampValue(ChokingAlgorithm::FixedSlots, ChokingAlgorithm::RateBased))
+#else
     , m_chokingAlgorithm(BITTORRENT_SESSION_KEY("ChokingAlgorithm"), ChokingAlgorithm::FixedSlots
         , clampValue(ChokingAlgorithm::FixedSlots, ChokingAlgorithm::RateBased))
+#endif
     , m_seedChokingAlgorithm(BITTORRENT_SESSION_KEY("SeedChokingAlgorithm"), SeedChokingAlgorithm::FastestUpload
```

Comment:
Now we have tweaked the `choking_algorithm` default value. Should we also tweak `seed_choking_algorithm` defaults?
AFAIK `FastestUpload` will only favor peers (who has fast download rate) and it doesn't have fairness in mind. Personally I would favor `anti_leech` being the default (if it works well with those notorious cheating clients) or at least `round_robin`.
What do you guys think?

Reply:
The seed_choking_algorithm default could be round_robin if it works correctly with <3 peer torrents.
There was an old bug that's probably already fixed...where round_robin needed at least 3 working peers before it would cycle upload slots through the peers correctly.
I've seen peers that refuse to download even when unchoked, which might also mess it up.

fastest_upload may favor peers that are uploading little and can download faster as a result, so it was only chosen by qBitTorrent as the stop-gap default due to the old bug with round_robin.

anti_leech doesn't just need to work well with notorious cheating clients that always report 0% to seeds...it needs to work well on marginal swarms as well. If Hit-And-Runners (via manual control or automated) are common on a public torrent, favoring upload to nearly complete peers is probably not the best default choice.

Even an automatic enabling of regular Super-Seeding might be worth considering on marginal swarms, which are painfully common on older/rarer public torrents.
Regular super-seeding works almost the exact opposite of anti_leech -- uploading to peers which have some but not typically all pieces.
Strict super-seeding is a bad default choice because it quits uploading until a peer reports it has a piece the seed previously uploaded to a different peer.

Reply:
IIRC cheating clients subvert the `anti-leech` choker. Between `round_robin` and `fastest_upload`, I'd rather keep the latter. I think in the long run it is worth it to maximize throughput. The faster we can distribute files to other peers, the faster they will distribute among themselves and to others, an so on.

Reply:
When there is excess peers, does fastest_upload mean never upload to slower peers?

So long as there's a steady stream of peers of hit-and-runners that allow only 1 (or 0) upload slot, set total upload speed <100 KB/sec, don't count overheads against speed limits, and also stop the torrent shortly after it reaches 100% ...what happens to the torrent swarm?

Reply:
@Seeker2 A majority of peers aren't like that... also that simply means slower peers will have to trade among themselves and eat the breadcrumbs thrown their way until the selfish fast peers leave the swarm. Everyone still completes the download anyway, as long as seeders stay.

Reply:
are fast peers less likely to share than slow peers? I imagine a model for this would be to upload to peers that will (eventually) have the highest share ratio

Reply:
@arvidn > are fast peers less likely to share than slow peers?

I don't think so. Fast peers tend to also have fast uplinks, and thus tend to share more.

Plus most people are not actively trying to hit and run or sabotage the swarm.

One must compare apples to apples. If if are thinking about "how much does a fast peer who immediately stops the torrent upon completion contribute to the swarm?", we must compare that to a slow peer in similar circumstances, i.e. "how much does a slow peer who immediately stops the torrent upon completion contribute to the swarm?". The former will most likely contribute more, because it most likely has faster uplink, and so for the short time it was sharing data, it share more than the slower peer.

Plus, I would speculate that faster peers with fiber connections and good routers/gateways tend to worry less about leaving a torrent seeding. A peer with slow uplink/downlink and network hardware that doesn't handle congestion very well is most likely to set ratio limits, up speed limits, etc.

Reply:
>also that simply means slower peers will have to trade among themselves and eat the breadcrumbs thrown their way via the optimistic unchokes until the selfish fast peers leave the swarm

I feel bad for the slower peers, this is almost like the pay-to-win concept in gaming...

>The former will most likely contribute more, because it most likely has faster uplink, and so for the short time it was sharing data, it share more than the slower peer.

I *imagine* this would be (more or less) tied to geographic location, faster peers would *usually* locate closer to each other (same or near continents) while people on the other side of globe would not be favored in this competition. IMO this *could* have impact on the data propagation rate (in the sense of geographically).

Also... since we are discussing just the default value, I think we should only consider whether the default option suits well with other default settings out of the box.

Reply:
In the sake of swarm fairness, maybe combine the two algorithms? ->

* Use "fastest_upload" for 50% of the slots (per torrent/global) and Round-Robin for the rest (slowest/idle 50% of the peers) 

* use different time-outs for "fastest" 50% group - higher timeout -  and slowest group 50% - lower timeout , maybe for idle peers - lowest timeout  (timeout = time-to RR a peer ) 

For experimentation have both the fastest %/split and timeouts - definable.

Reply:
@Chocobo1 
> I feel bad for the slower peers, this is almost like the pay-to-win concept in gaming...

I don't think it is that bad. Perhaps I painted an excessively grim picture. I think "trickle down" works nicely for the torrent economy - it is overall best for everyone to prioritize the fastest peers. This is just "educated" speculation, I have no hard data to back this up. Ultimately there are no miracles. A peer based in US/Australia will most likely download slowly, that's life. You get a better Internet experience overall if you pay more and live in a country with good infrastructure.

> I _imagine_ this would be (more or less) tied to geographic location, faster peers would _usually_ locate closer to each other (on the same or near continents) while people on the other side of globe would not be favored in this competition. IMO this _might_ have impact on the data propagation rate (in the sense of geographically).

Peers that are geographically close might be the first ones that get connected to, but eventually everyone gets a shot thanks to the optimistic unchoke, no?

> Also... since we are discussing just the default value, I think we should only consider whether the default option suits well with other out of the box settings.

I think `fastest_upload` has always been a good out-of-the-box default. I think the extent to which it _might_ make swarms less accommodating for slower peers at the beginning (and slower peers should have no expectations of racing anyway...) is worth it compared to the increased uplink saturation it allows. I'd be concerned about user complains that "qbittorrent doesn't seed as much" if we change it.

Reply:
>You get a better Internet experience overall if you pay more and live in a country with good infrastructure.

IMO you pay more to get better service from your ISP/country and that is perfectly normal, but that doesn't mean you would be VIP in a file sharing protocol, that is two different things.

>Peers that are geographically close might be the first ones that get connected to, but eventually everyone gets a shot thanks to the optimistic unchoke, no?

AFAIK optimistic unchoke is only about trying new peers, it doesn't help much in terms of distributing fairly when favoring faster peers.

>I'd be concerned about user complains that "qbittorrent doesn't seed as much" if we change it.

I sense that there are people expecting to win the upload race or whatever that is, while I can't really say that is right or wrong, this expectation is really affecting the default distribution behavior in some way.
Do people really expect to win that race with a client with out of the box settings?

Another related question, when using "rate based" choking algorithm with "round-robin" algorithm, assuming uploading to all peers are slow, wouldn't it open more upload slots in order to utilize the unused bandwidth? If so, I imagine this would spread the data more evenly to all peers than just only the fastest peers, I mean this lead to a better distributing behavior isn't?

I guess there isn't much sound on changing the default, then that's that.

Reply:
Opening up more slots and uploading evenly to more peers is actually a bad idea. Every peer wants a different chunk. The more slots I open, the more pieces I have to read out from disk, hash them, store to cache. Whereas if I open up less number of faster slots I utilise less of my other resources and upload faster as well. This may sound unfair but in reality it isn’t. It’s actually best for the entire swarm. The fast downloading peers would eventually serve as seeders and will upload to the slow peers eventually. So it’s a win win situation.

Reply:
I think you're right from the viewpoint of performance... Seems real P2P is another utopia, there will always be hierarchy in some form, never expected I would ever see it this way, cool.

Reply:
@Chocobo1 
> IMO you pay more to get better service from your ISP/country and that is perfectly normal, but that doesn't mean you would be VIP in a file sharing protocol, that is two different things.

If you are a fast peer, it makes sense that you would be sort of VIP, because of this:

>  It’s actually best for the entire swarm. The fast downloading peers would eventually serve as seeders and will upload to the slow peers eventually. So it’s a win win situation.

This is what I meant with "trickle down", @an0n666 said it better.

> AFAIK optimistic unchoke is only about trying new peers, it doesn't help much in terms of distributing fairly when favoring faster peers.

Yeah, but I thought your point was about geographically distributing the data. Thanks to the optimistic unchoke, even a "distant" peer that only connected at a later time will be unchoked eventually. No matter where a peer is, they will be unchoked if they are faster, speed is the real differentiator.

> I sense that there are people expecting to win the upload race or whatever that is, while I can't really say that is right or wrong, this expectation is really affecting the default distribution behavior in some way.
> Do people really expect to win that race with a client with out of the box settings?

Here I was not thinking of racing (qBittorrent can't even race right now because it lacks the "fast initial announce retry" feature) - I'm just thinking of the average user who would consider qBittorrent "not the best for seeding". Again, I really don't think `fastest_upload` is hurting swarm health in any meaningful way. Whereas using something other that `fastest_upload` has the potential to make a huge difference for each user individually. Put another way, I think it's better to decrease swarm health by 0.5% if that improves potential user seeding performance by 5% (or more). And this is assuming there is even a hit to swarm health at all.

> Another related question, when using "rate based" choking algorithm with "round-robin" algorithm, assuming uploading to all peers are slow, wouldn't it open more upload slots in order to utilize the unused bandwidth? If so, I imagine this would spread the data more evenly to all peers than just only to the fastest peers, I mean this lead to a better distributing behavior isn't?

`rate_based` applies to peers you are downloading from, while `round_robin` is for peers you are seeding to. Assuming the global slots limit is not being hit by the unchokes due to `rate_based`, the remaining slots will be used to unchoke peers in a `round_robin` fashion to seed to. This is fairer on the surface. However, this will mean your upload capacity is wasted  and it will hurt swarm performance, since a lot of the time you could be cycling through slow peers that do not saturate you. It would be more efficient to focus on uploading to the faster peers first, because then they can also help to distribute the data to the slower peers, and so on (this is the "trickle down" argument).

> I guess there isn't much sound on changing the default, then that's that.

No news is good news - I haven't seen issues complaining about this, so it's probably fine. All of qBittorrent's issues with seeding that get reported now and then have other causes.

> Seems real P2P is another utopia

I wouldn't say that, it's a bit pessimistic :). We already have real P2P, we are just trying to optimize it given the constraints of finite, non-uniform network bandwidth and storage medium I/O performance of the swarm.

Reply:
"You get a better Internet experience overall if you pay more and live in a country with good infrastructure."
"A peer based in US/Australia will most likely download slowly, that's life."
"faster peers would usually locate closer to each other (on the same or near continents)"

Not really.
High prices have more to do with price-gouging and LACK of competition.
I should know, I pay too much for a ComCast connection to the internet.
Much of online gaming I play is on servers hosted in USA or Canada.
Many of the websites I visit are hosted in the USA or at least mirrored in the USA.
My connectivity to Europe and Japan is good, with higher pingtimes due almost entirely to distance rather than bandwidth bottlenecks and packet loss.

Many of the fastest peers I see are far away -- Sweden, Finland, Japan, South Korea...
And there are university-level BitTorrent studies to back this up.

Reply:
You always bring up super seeding but you don’t talk about the downsides of using it. If the peers I’m uploading to are all firewalled then super seeding will fail. If peers limit their upload speed it’d fail as well even if not firewalled. For it to work properly every peer must be reachable and sharing. Which is impossible to achieve in most swarms.

Reply:
With normal super seeding, uploading of rare pieces continues even if NO peers show as having pieces already uploaded.
Only strict super seeding quits uploading if it doesn't see pieces elsewhere, and I do not recommend strict for EXACTLY the reasons you stated!

Normal super seeding will even use as many upload slots as allowed, instead of letting some go idle because peers don't appear to be sharing (possibly because of firewalls or XunLei).

Even firewalled peers isn't a total disaster so long as the torrent is public and most everyone is using uTP and PEX (both enabled by default).
Then, uTP can NAT hole-punch so 2 firewall peers can share directly with each other.

Reply:
Libtorrent uses UPnP and NAT-PMP. There’s no NAT hole punching.

Reply:
Please do some testing of that.
If the uTP NAT STUN is broken, make an issue post on the subject or comment on an existing one.
Found one here:
https://github.com/arvidn/libtorrent/issues/604

Reply:
technically it's not STUN, the hole punching used in bittorrent is decentralized

Reply:
IMO the _strict_ superseeding setting should be outright removed from qBIttorrent. Some users may activate it because of those YT tutorials or some redditor, or because the name itself is kind of misleading - "strict" and "super": that means qBIttorrent is going to seed like a boss, right?

In reality, _strict_ superseeding will lead to catastrophic failures to seed, possibly leading to the death of the torrent, in far too many circumstances to be worth it. It is dangerous in smaller swarms, which is exactly what we don't want.

As far as _non strict_ superseeding goes, from what I can see in the libtorrent docs, it can't be enabled in the settings directly, it is a `torrent_flags_t`. @arvidn what would be the pros/cons of generally enabling it? I'm not quite sure about the differences vs. the _strict_ version. Can it lead to catastrophic seeding failure in any circumstance? If so, it is also not worth it.

Reply:
In private trackers I’ve seen literally hundreds of users using the strict super seeding more. This is because the name is misleading. There’s nothing super in super seeding. They presume this to increase their upload rate.

Reply:
as a historical note. the original super seeding feature (from one of the early python clients) implemented the strict behavior. So in a way, that's "normal super seeding". Also documented in [BEP 16](https://www.bittorrent.org/beps/bep_0016.html)

Reply:
@arvidn Ok, for disambiguation let's call normal super seeding "strict super seeding" and the other variant "non-strict super seeding". I still think the _strict_ variant should be removed from qBittorrent as an available option.

Plus, given that it is such a dangerous option to use, why does libtorrent still support it? I'd really like to hear the technical reasoning about the potential advantages of both variants. Maybe one or both of them were relevant in 2008 but not anymore. The whole BEP is about tricking other peers, which seems like malicious/anti-social behavior.

Reply:
I'm not necessarily sold on the merits of *any* super seeding. The reason libtorrent supports it is because it was requested once, and never removed.

The main scenario where people might use it is if:

1. they are the *initial* seed (once there is one other seed in the swarm, the strict super seeding is pointless, but the relaxed one still has an effect).
2. their upload bandwidth is metered, and sending fewer bytes is better, even if it means everything will be slower.

Maybe the option should be renamed to `BEP 16` instead. That way people probably won't enable it because they like the sound of the name

Reply:
The strict super seeding option should be gone from qBt. Almost everyone uses it without knowing the consequences. However I don’t have much against the relaxed one for specific torrents.

Reply:
@arvidn

> I'm not necessarily sold on the merits of _any_ super seeding. The reason libtorrent supports it is because it was requested once, and never removed.
> 
> The main scenario where people might use it is if:
> 
>     1. they are the _initial_ seed (once there is one other seed in the swarm, the strict super seeding is pointless, but the relaxed one still has an effect).

Can you elaborate more on the additional effect of the non-strict version?

As for the moment where the initial seed is the only seed, is the advantage of the strict variant significant during that time?

>     2. their upload bandwidth is metered, and sending fewer bytes is better, even if it means everything will be slower.

I don't think we should explicitly cater to this. Users with metered connections should set upload speed limits in qBittorrent (or data limits if that is ever implemented) or speed/data limits at the OS level.

> Maybe the option should be renamed to `BEP 16` instead. That way people probably won't enable it because they like the sound of the name

Good idea, or maybe even `BEP 16 (use with extreme caution)`, optionally with a link to a wiki page clearly outlining the pros/cons, with the general tone of "if you have to ask, then you don't need it".

Reply:
I agree strict super seeding should be removed as it causes more harm than good.
Pure leeching BitTorrent clients were more common percent-wise all over the world (outside of China, that is) back in 2008. Strict super seeding was even recommended for people with low upload who were the 1st-and-only seed on a torrent -- they would (in theory!) waste little upload on "bad" peers. So strict super-seeding not uploading at all when pieces aren't seen was a feature not a bug.
Considering HAVE message screw-ups and not reporting percentage complete (especially to seeds, to save bandwidth) were more common back then, it never made much sense.

As for name/meaning confusion over super seeding, because as others point out it's not terribly "super"...remove the word Super and either call it either:
1. Strict Seeding
2. Initial Seeding
3. Early Seeding (to prevent confusion with uTorrent's broken Initial Seeding method)
...worth putting to a vote/poll? Anyone else got any better names for it?

I'm in favor of keeping the "softer" less-drastic form of super seeding/initial seeding because without it...
If a seed is "hit" by a huge number of new peers all wanting 1st/last pieces or sequentially downloading...it will repeatedly comply even if the seed is the only source of rare pieces.
Seeds are supposed to be the go-to for rare pieces that peers don't have.
HAVE FAST could encourage better behavior for all, but almost nothing seems to work with that.

Reply:
@FranciscoPombal really the whole point of super seeding is to prioritize uploading as little as possible over performance. The strict mode only works if you're the initial seed because it *enforces* that peers it sends to share the piece. It will wait with giving you another piece until it can see at least one other peer with the last you sent. So, as you can imagine, you stall the pipe in order to enforce this.

The less strict super seeding is really just a way for the seeder to pick pieces, as opposed to the downloader, which normally decides what to pick. So it's up to the seeder to decide what's rare and not, and it will send various rare pieces to peers. This mode works even if there are other seeds in the swarm as well, since it just inverts the control of what peers download.

Reply:
@arvidn
> it just inverts the control of what peers download

So the non-strict version is just a another way of doing the same thing as the protocol does otherwise, but in a different way, without any advantage? In that case I would consider it to be bloat.

@Seeker2

I'm in favor of keeping the "softer" less-drastic form of super seeding/initial seeding because without it...
If a seed is "hit" by a huge number of new peers all wanting 1st/last pieces or sequentially downloading...it will repeatedly comply even if the seed is the only source of rare pieces.

That seems like an unrealistic scenario. And even if literally everyone in the swarm was downloading sequentially, what's the point of starting to seed a torrent if you're not going to be able to stay online for long enough to seed at least one copy. If you don't want to spread bandwidth to thin, just limit the number of upload slots.